text/0617-cloudfront-oac-l2.md

```diff
@@ -0,0 +1,877 @@
+   */
+  NEVER = 'never',
+  /**
+   * Sign origin requests only if the viewer request
+   * doesn't contain the Authorization header.
+   */
+  NO_OVERRIDE = 'no-override',
+}
+
+/**
+ * The signing protocol of the origin access control.
+ */
+export enum SigningProtocol {
+  /**
+   * The AWS Signature Version 4 signing protocol.
+   */
+  SIGV4 = 'sigv4',
+}
+
+/**
+ * Options for how CloudFront signs requests.
+ */
+export class Signing {
+  /**
+   * Sign all origin requests using the AWS Signature Version 4 signing protocol.
+   */
+  public static readonly SIGV4_ALWAYS = new Signing(SigningProtocol.SIGV4, SigningBehavior.ALWAYS);
+  /**
+   * Do not sign any origin requests.
+   */
+  public static readonly SIGV4_NEVER = new Signing(SigningProtocol.SIGV4, SigningBehavior.NEVER);
+  /**
+   * Sign only if the viewer request doesn't contain the Authorization header
+   * using the AWS Signature Version 4 signing protocol.
+   */
+  public static readonly SIGV4_NO_OVERRIDE = new Signing(SigningProtocol.SIGV4, SigningBehavior.NO_OVERRIDE);
+
+  /**
+   * The signing protocol
+   */
+  public readonly protocol: SigningProtocol;
+  /**
+   * Which requests CloudFront signs.
+   */
+  public readonly behavior: SigningBehavior;
+
+  public constructor(protocol: SigningProtocol, behavior: SigningBehavior) {
+    this.protocol = protocol;
+    this.behavior = behavior;
+  }
+}
+
+/**
+ * An Origin Access Control.
+ * @resource AWS::CloudFront::OriginAccessControl
+ * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-cloudfront-originaccesscontrol.html
+ */
+export class OriginAccessControl extends Resource implements IOriginAccessControl {
+  /**
+   * Imports an origin access control from its id and origin type.
+   */
+  public static fromOriginAccessControlAttributes(scope: Construct, id: string, attrs: OriginAccessControlAttributes): IOriginAccessControl {
+    class Import extends Resource implements IOriginAccessControl {
+      public readonly originAccessControlId = attrs.originAccessControlId;
+      public readonly originAccessControlOriginType = attrs.originAccessControlOriginType;
+    }
+    return new Import(scope, id);
+  }
+  /**
+   * Creates an origin access control for a S3 origin.
+   */
+  public static forS3(scope: Construct, id: string, props?: CommonOriginAccessControlProps): IOriginAccessControl {
+    return new OriginAccessControl(scope, id, {
+      ...props,
+      originAccessControlOriginType: OriginAccessControlOriginType.S3,
+    });
+  }
+
+  public readonly originAccessControlId: string;
+  public readonly originAccessControlOriginType: string;
+
+  constructor(scope: Construct, id: string, props: OriginAccessControlProps) {
+    super(scope, id);
+    this.originAccessControlOriginType = props.originAccessControlOriginType;
+
+    // Check if origin access control name is 64 characters or less
+    if (props.originAccessControlName) {
```

Comment:
This is a larger topic, but maybe worth discussing a bit here. It's very standard in CDK to validate these things at synth time, since that's a way better experience than CFN validation. I think we should continue to do this, especially when there's no chance of a regression. We can always loosen these checks if it turns out that we got them wrong in the initial implementation.

If we omit them here, somebody later is going to come submit a PR to add it, and it's much slower to validate that the check is correct then than it is when we're the authors.